app/com/linkedin/drelephant/tony/fetchers/TonyFetcher.java

```diff
@@ -0,0 +1,90 @@
+import com.linkedin.tony.Constants;
+import com.linkedin.tony.TonyConfigurationKeys;
+import com.linkedin.tony.events.Event;
+import com.linkedin.tony.util.ParserUtils;
+import java.io.IOException;
+import java.util.Date;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.Logger;
+
+
+public class TonyFetcher implements ElephantFetcher<TonyApplicationData> {
+  private static final Logger _LOGGER = Logger.getLogger(TonyFetcher.class);
+  private final Path _finishedDir;
+  private final FileSystem _fs;
+
+  /**
+   * Constructor for {@link TonyFetcher}.
+   * @param fetcherConfig  the fetcher configuration
+   * @throws IOException
+   */
+  public TonyFetcher(FetcherConfigurationData fetcherConfig) throws IOException {
+    Configuration conf = new Configuration();
+
+    String tonyConfDir = System.getenv(Constants.TONY_CONF_DIR);
+    if (fetcherConfig.getParamMap().containsKey(Constants.TONY_CONF_DIR)) {
+      tonyConfDir = fetcherConfig.getParamMap().get(Constants.TONY_CONF_DIR);
+    }
+    _LOGGER.info("Using TonY conf dir: " + tonyConfDir);
+
+    conf.addResource(new Path(tonyConfDir + Path.SEPARATOR + Constants.TONY_SITE_CONF));
+    _finishedDir = new Path(conf.get(TonyConfigurationKeys.TONY_HISTORY_FINISHED));
+    _fs = _finishedDir.getFileSystem(conf);
+  }
+
+  @Override
+  public TonyApplicationData fetchData(AnalyticJob job) throws Exception {
+    _LOGGER.debug("Fetching data for job " + job.getAppId());
+    long finishTime = job.getFinishTime();
+    Date date = new Date(finishTime);
+
+    // TODO: We are deriving yyyy/MM/dd from the RM's application finish time, but the TonY Portal actually creates the
+    // yyyy/MM/dd directory based off the end time embedded in the jhist file name. This end time is taken before the
+    // application finishes and thus is slightly before the RM's application finish time. So it's possible that the
+    // yyyy/MM/dd derived from the RM's application finish time is a day later and we may not find the history files.
+    // In case we don't find the history files in yyyy/MM/dd, we should check the previous day as well.
+    String yearMonthDay = ParserUtils.getYearMonthDayDirectory(date);
+    Path jobDir = new Path(_finishedDir, yearMonthDay + Path.SEPARATOR + job.getAppId());
+    _LOGGER.debug("Job directory for " + job.getAppId() + ": " + jobDir);
+
+    // parse config
+    Path confFile = new Path(jobDir, Constants.TONY_FINAL_XML);
+    if (!_fs.exists(confFile)) {
+      // for backward compatibility, see https://github.com/linkedin/TonY/issues/271
+      confFile = new Path(jobDir, "config.xml");
+    }
+    Configuration conf = new Configuration(false);
+    if (_fs.exists(confFile)) {
+      conf.addResource(_fs.open(confFile));
+    }
+
+    // parse events
+    List<Event> events = ParserUtils.parseEvents(_fs, jobDir);
```

Comment:
What kind of events are returned for each task ? Can there be multiple events like "TASK_STARTED" and "TASK_FINISHED" for same task. Will be good if you can give some context here. Also explain what kind of metrics are you expecting for each event.

Reply:
Each task should only have at most a single `TASK_STARTED` and `TASK_FINISHED` event. I added a link to https://github.com/linkedin/TonY/blob/master/tony-core/src/main/avro/EventType.avsc which contains a list of all the event types.

I also added some comments about what information we're extracting from the `TASK_STARTED` and `TASK_FINISHED` events.